internal/backend/local/backend_apply.go

```diff
@@ -266,80 +271,92 @@ func (b *Local) opApply(
-					// its value on to the ApplyOpts.
-					applyTimeValues[varName] = val
-				} else {
-					// TODO: We should probably actually tolerate this if the new
-					// value is equal to the value that was saved in the plan, since
-					// that'd make it possible to, for example, reuse a .tfvars file
-					// containing a mixture of ephemeral and non-ephemeral definitions
-					// during the apply phase, rather than having to split ephemeral
-					// and non-ephemeral definitions into separate files. For initial
-					// experiment we'll keep things a little simpler, though, and
-					// just skip this check if we're doing a combined plan/apply where
-					// the apply phase will therefore always have exactly the same
-					// inputs as the plan phase.
+				// If this is an apply-time variable, the user must supply a
+				// value during apply: it can't be null.
+				if applyTimeVar && val.Value.IsNull() {
+					diags = diags.Append(&hcl.Diagnostic{
+						Severity: hcl.DiagError,
+						Summary:  "Ephemeral variable must be set for apply",
+						Detail: fmt.Sprintf(
+							"The ephemeral input variable %q was set during the planning phase, and so must be set again during the apply phase.",
+							varName,
+						),
+					})
+					continue
+				}
 
+				// If we get here, we are in possession of a non-null
+				// ephemeral apply-time input variable, and need only pass
+				// its value on to the ApplyOpts.
+				applyTimeValues[varName] = val
+			} else {
+				// If a non-ephemeral variable is set differently between plan and apply, we should emit a diagnostic.
+				plannedVariableValue, ok := plan.VariableValues[varName]
+				if !ok {
 					diags = diags.Append(&hcl.Diagnostic{
 						Severity: hcl.DiagError,
 						Summary:  "Can't set variable when applying a saved plan",
-						Detail:   fmt.Sprintf("The variable %s cannot be set using the -var and -var-file options when applying a saved plan file, because a saved plan includes the variable values that were set when it was created. To declare an ephemeral variable which is not saved in the plan file, use ephemeral = true.", varName),
+						Detail:   fmt.Sprintf("The variable %s cannot be set using the -var and -var-file options when applying a saved plan file, because it is neither ephemeral nor has it been declared during the plan operation. To declare an ephemeral variable which is not saved in the plan file, use ephemeral = true.", varName),
 						Subject:  rng,
 					})
+					continue
 				}
 
+				val, err := plannedVariableValue.Decode(cty.DynamicPseudoType)
+				if err != nil {
+					diags = diags.Append(&hcl.Diagnostic{
+						Severity: hcl.DiagError,
+						Summary:  "Could not decode variable value from plan",
+						Detail:   fmt.Sprintf("The variable %s could not be decoded from the plan. %s. This is a bug in Terraform, please report it.", varName, err),
+						Subject:  rng,
+					})
+				} else {
+					if v.Value.Equals(val) == cty.False {
```

Comment:
`cty.Values` can't be directly compared (which is why you're using `Equals`). `cty.Value` has `True()` and `False()` methods for boolean operations.

Reply:
Note this probably works fine here due to the context so we don't ned to deal with nulls, unknowns, marks, etc. it's just that comparing a boolean in a boolean expression just looks so wrong ;)

Reply:
Ah, I didn't know about `.False()` and `.True()`, TIL!